jicoco-kotlin/src/main/kotlin/org/jitsi/health/AbstractHealthCheckService.kt

```diff
@@ -0,0 +1,190 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jitsi.health
+
+import org.jitsi.utils.concurrent.PeriodicRunnable
+import org.jitsi.utils.concurrent.RecurringRunnableExecutor
+import org.jitsi.utils.logging2.Logger
+import org.jitsi.utils.logging2.LoggerImpl
+import org.osgi.framework.BundleActivator
+import org.osgi.framework.BundleContext
+import java.time.Clock
+import java.time.Duration
+import java.time.Instant
+
+abstract class AbstractHealthCheckService(
+    val interval: Duration = Duration.ofSeconds(10),
+    /**
+     * If no health checks have been performed in the last {@code timeout}
+     * period, the service is considered unhealthy.
+     */
+    val timeout: Duration = Duration.ofSeconds(30),
+    /**
+     * The maximum duration that a call to {@link #performCheck()} is allowed
+     * to take. If a call takes longer, the service is considered unhealthy. A
+     * value of {@code null} indicates no max check duration.
+     * <p>
+     * Note that if a check never completes, we rely on {@link #timeout} instead.
+     */
+    private val maxCheckDuration: Duration = Duration.ofSeconds(3),
+    /**
+     * If set, a single health check failure after the initial
+     * {@link #STICKY_FAILURES_GRACE_PERIOD}, will be result in the service
+     * being permanently unhealthy.
+     */
+    private val stickyFailures: Boolean = false,
+    /**
+     * Failures in this period (since the start of the service) are not sticky.
+     */
+    private val stickyFailuresGracePeriod: Duration = stickyFailuresGracePeriodDefault
+): BundleActivator, HealthCheckService, PeriodicRunnable(interval.toMillis())
+{
+    @JvmOverloads
+    constructor(interval: Duration, timeout: Duration, maxCheckDuration: Duration, stickyFailures: Boolean) : this(
+        interval = interval,
+        timeout = timeout,
+        maxCheckDuration = maxCheckDuration,
+        stickyFailures = stickyFailures,
+        stickyFailuresGracePeriod = stickyFailuresGracePeriodDefault)
+
+    private val logger: Logger = LoggerImpl(javaClass.name)
+
+    /**
+     * The executor which runs {@link #performCheck()} periodically.
+     */
+    private var executor: RecurringRunnableExecutor? = null
+
+    /**
+     * The exception resulting from the last health check. When the health
+     * check is successful, this is {@code null}.
+     */
+    private var lastResult: Exception? = null
+
+    /**
+     * The time the last health check finished being performed.
+     */
+    private var lastResultTime = Instant.MIN
+
+    /**
+     * The time when this service was started.
+     */
+    private var serviceStartTime = Instant.MAX
+
+    /**
+     * Whether we've seen a health check failure (after the grace period).
+     */
+    private var hasFailed = false
+
+    private val clock = Clock.systemUTC()
```

Comment:
Can you take the clock as a param?  Easier to test things if it can be overridden.